src/openslide-vendor-philips-isyntax.c

```diff
@@ -0,0 +1,397 @@
+    }
+
+    // draw it
+    g_autoptr(cairo_surface_t) surface =
+            cairo_image_surface_create_for_data((unsigned char *) tiledata,
+                                                CAIRO_FORMAT_ARGB32,
+                                                tw, th, tw * 4);
+    cairo_set_source_surface(cr, surface, 0, 0);
+    // https://lists.cairographics.org/archives/cairo/2012-June/023206.html
+    // Those are the operators that are observed to work:
+    //    w CAIRO_OPERATOR_SATURATE (current_cairo_operator, aka default OpenSlide)
+    //    w CAIRO_OPERATOR_OVER ("This operator is cairo's default operator."),
+    //    w CAIRO_OPERATOR_DEST_OVER,
+    // SATURATE takes ~12sec to read a dummy slide (forcing all tiles to not exist), OVER & DEST_OVER take 3.5 sec
+    // for same setup. Selecting OVER as the Cairo's default operator, the three outputs are identical.
+    cairo_operator_t current_cairo_operator = cairo_get_operator(cr);
+    cairo_set_operator(cr, CAIRO_OPERATOR_OVER);
+    cairo_paint(cr);
+    cairo_set_operator(cr, current_cairo_operator);
+    return true;
+}
+
+static void add_float_property(openslide_t *osr, const char* property_name, float value) {
+    g_hash_table_insert(osr->properties, g_strdup(property_name),
+                        _openslide_format_double(value));
+}
+
+static bool philips_isyntax_open(
+        openslide_t *osr,
+        const char *filename,
+        struct _openslide_tifflike *tl G_GNUC_UNUSED,
+        struct _openslide_hash *quickhash1 G_GNUC_UNUSED,
+        GError **err) {
+    // Do not allow multithreading in opening.
+    // https://docs.gtk.org/glib/method.Mutex.init.html:
+    //   "It is not necessary to initialize a mutex that has been statically allocated."
+    static GMutex static_open_mutex;
+    g_mutex_lock(&static_open_mutex);
+    static bool threadmemory_initialized = false;
+    if (!threadmemory_initialized) {
+        // TODO(avirodov): review libisyntax_init() thread safety, if thread safe, remove the mutex here.
+        libisyntax_init();
+        threadmemory_initialized = true;
+    }
+    g_mutex_unlock(&static_open_mutex);
+    LOG("Opening file %s", filename);
+
+    philips_isyntax_t* data = malloc(sizeof(philips_isyntax_t));
+
+    osr->data = data;
+    isyntax_error_t open_result = libisyntax_open(filename, /*is_init_allocators=*/0, &data->isyntax);
+    LOG_VAR("%d", (int)open_result);
+    // LOG_VAR("%d", data->isyntax->image_count); // TODO(avirodov): getter.
+    if (open_result != LIBISYNTAX_OK) {
+        free(data);
+        g_prefix_error(err, "Can't open file.");
+        return false;
+    }
+
+    // Initialize the cache (global, if requested).
+    bool is_global_cache = true;
+    int cache_size = 2000;
+    const char* str_is_global_cache = g_environ_getenv(g_get_environ(), "OPENSLIDE_ISYNTAX_GLOBAL_CACHE");
+    const char* str_cache_size = g_environ_getenv(g_get_environ(), "OPENSLIDE_ISYNTAX_CACHE_SIZE");
```

Comment:
TLDR: let's leave them in please, at least for initial tuning.

>  How important are these tunables?

Not sure myself. Ideally cache_size = O(available_mem / num_processes). But we don't know either in an ML scenario, not from within libisyntax. So I expect this will require some tuning by the user to get best usage of available memory - which is important for avoiding re-reading dependent tiles (very slow disk ops, very many tile dependencies).

However, currently we blow out of memory anyway on some specific ML scenarios, and tuning cache_size doesn't help (because the issue is elsewhere I think), so I don't actually use it. I will keep it for myself for tuning in the short term anyway (even if in my fork only).

I do expect it to be useful long-term because of the unknowns in the equation above.

> try to piggyback on openslide_set_cache().
Note that these are two different caches. Sure, we can use the shared OpenSlide cache as a hint, but I think there's no downside to having a shared cache always, as long as you can set cache size. In viewer scenario (one slide open), the user would set cache to MAX_INT and sharing doesn't matter. In ML scenario, more careful tuning is needed, but you'd want shared cache for sure if you want to keep 100 or 1000 isyntax files open in parallel, reading a handful of tiles from each.